### PR TITLE
Find coordinates given parcellations image

### LINKED ISSUE #646: Add functions to grab center of mass coords of probabilistic and parcellation atlases
would simplify connectome plots.

### PR DESCRIPTION
**Pull request by KamalakerDadi**
Fixes #646 , #1111  and supersedes #664 

Things to do:

- Improving documentation of the functions ```find_parcellation_cut_coords``` and ```find_probabilistic_atlas_cut_coords```
- Improving the example
- Fixing tests and adding more tests

### CODE REVIEW

**Comment by bthirion** [comment]
Looks good besides small details. Thx.

**Comment by KamalakerDadi** [comment]
All environments are fixed for review. Can I have some reviews please ?
Thanks

File: nilearn/plotting/tests/test_find_cuts.py [hunk 1]
```diff
@@ -156,3 +158,98 @@ def test_fast_abs_percentile_no_index_error_find_cuts():
...
+                               (x_map_b / 2., y_map_b / 3., z_map_b / 4.),
+                               rtol=6e-2)
+    np.testing.assert_allclose((coords[2][0], coords[2][1], coords[2][2]),
+                               (x_map_c / 2., y_map_c / 3., z_map_c / 4.),
+                               rtol=6e-2)
+
+
+def test_find_probabilistic_atlas_cut_coords():
+    # make data
+    arr1 = np.zeros((100, 100, 100))
```
Comment (bthirion) [inline nilearn/plotting/tests/test_find_cuts.py]: Do you need to take such lrage images ? doesn't it take too much time ?

File: nilearn/plotting/tests/test_find_cuts.py [hunk 2]
```diff
@@ -156,3 +158,98 @@ def test_fast_abs_percentile_no_index_error_find_cuts():
...
+    x_map_a, y_map_a, z_map_a = 30, 40, 50
+    arr1[x_map_a - 10:x_map_a + 10, y_map_a - 20:y_map_a + 20, z_map_a - 30: z_map_a + 30] = 1
+
+    arr2 = np.zeros((100, 100, 100))
+    x_map_b, y_map_b, z_map_b = 40, 50, 60
+    arr2[x_map_b - 10:x_map_b + 10, y_map_b - 20:y_map_b + 20, z_map_b - 30: z_map_b + 30] = 1
+
+    data = np.concatenate((arr1[..., np.newaxis], arr2[..., np.newaxis]), axis=3)
```
Comment (bthirion) [inline nilearn/plotting/tests/test_find_cuts.py]: you might include an empty map inbetween arr1 and arr2

File: nilearn/plotting/tests/test_find_cuts.py (hunk 1 above)
Comment (KamalakerDadi) [reply nilearn/plotting/tests/test_find_cuts.py]: Not much time. I need a bit of large images to grab the centroids of the non-overlapping blobs. Just to pass the tests with high tolerance in ```np.assert_allclose```.

**Comment by KamalakerDadi** [comment]
Do you have any objection on the naming of the functions ?

The one with labels image is named as ```find_parcellations_cut_coords``` and one with probabilistic atlases are named as ```find_probabilistic_atlas_cut_coords```.

**Comment by bthirion** [comment]
Maybe `find_probabilistic_atlas_cut_coords` and `find_atlas_cut_coords` for the sake of consistency.

File: examples/03_connectivity/plot_atlas_comparison.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+The time series of all subjects of the ADHD Dataset are concatenated to create
+parcel-wise correlation matrices for each atlas.
+
+# author: Amadeus Kanaan
+
+"""
+import numpy as np
+from nilearn import datasets
+from nilearn.input_data import NiftiLabelsMasker
+from nilearn import plotting
```
Comment (GaelVaroquaux) [inline examples/03_connectivity/plot_atlas_comparison.py]: Move the imports in the cells where they are used.

File: examples/03_connectivity/plot_atlas_comparison.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+print('Destrieux atlas nifti image (3D) is located at: %s' % destrieux['maps'])
+print('Yeo atlas nifti image (3D) with 17 parcels and liberal mask is located '
+      'at: %s' % yeo['thick_17'])
+print('Harvard Oxford atlas nifti image (3D) thresholded at .25 is located '
+      'at: %s' % harvard_oxford['maps'])
+# Store atlases and thresholds in a dictionary. Atlases are for extracting
+# coordinates and threshold for displaying the strength of the connectivity.
+atlases_and_thresholds = {'Destrieux Atlas (struct)': [destrieux['maps'], '97%'],
+                          'Yeo Atlas 17 thick (func)': [yeo['thick_17'], '70%'],
+                          'Harvard Oxford > 25% (struct)': [harvard_oxford['maps'], '90%']}
```
Comment (GaelVaroquaux) [inline examples/03_connectivity/plot_atlas_comparison.py]: I think that we can use 80% threshold for everyone, and simplify things.

File: examples/03_connectivity/plot_atlas_comparison.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
+"""
+Extracting resting-state signals from different atlases for comparison
+======================================================================
+
+In this example :class:`nilearn.input_data.NiftiLabelsMasker` is used to
+extract time series from nifti objects using different parcellation atlases.
```
Comment (GaelVaroquaux) [inline examples/03_connectivity/plot_atlas_comparison.py]: An important part of the example is to turn a parcellation into a connectome for visualization. This requires choosing centers of each parcel via  plotting.find_parcellation_cut_coords.

File: examples/03_connectivity/plot_atlas_comparison.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
+"""
+Extracting resting-state signals from different atlases for comparison
```
Comment (GaelVaroquaux) [inline examples/03_connectivity/plot_atlas_comparison.py]: Comparing connectomes on different reference parcellation atlases

File: examples/03_connectivity/plot_atlas_comparison.py (hunk 5 above)
Comment (GaelVaroquaux) [reply examples/03_connectivity/plot_atlas_comparison.py]: We should demo also find_probabilistic_atlas_cut_coords in the same example, in a second section. Hence, we should rephrase the above as "An important part of the example is to turn an atlas into a connectome for visualization. The requires choosing centers for each parcel or network, via plotting.find_parcellation_cut_coords or plotting.find_probabilistic_atlas_cut_coords depending on whether the atlas is a parcellation or a set of maps.

File: examples/03_connectivity/plot_atlas_comparison.py (hunk 6 above)
Comment (GaelVaroquaux) [reply examples/03_connectivity/plot_atlas_comparison.py]: Given that we are going to demo both functions in this example, remove parcellation from above.

File: examples/03_connectivity/plot_atlas_comparison.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    # calculate correlation matrix and display
+    correlation_matrix = np.corrcoef(time_series.T)
+
+    # grab center coordinates for atlas labels
+    coordinates = plotting.find_parcellation_cut_coords(atlas)
+
+    # plot connectome
+    plotting.plot_connectome(correlation_matrix, coordinates,
+                             edge_threshold=threshold, title=name)
+
```
Comment (GaelVaroquaux) [inline examples/03_connectivity/plot_atlas_comparison.py]: We need a second section demoing probabilistic atlases with the harvard oxford cortical (split), the msdl and the smith RSN.

File: nilearn/plotting/find_cuts.py [hunk 8]
```diff
@@ -323,3 +323,127 @@ def find_cut_slices(img, direction='z', n_cuts=7, spacing='auto'):
...
+
+        # Case 1: One label in one hemisphere
+        if np.all(left_hemi == False) or np.all(right_hemi == False):
+            pass
+        # Case 2: Connected component spanning both hemispheres
+        # elif np.any(right_hemi[int(x)]) and np.any(left_hemi[int(x-1)]):
+        #    if not np.any(right_hemi[int(x-10)]) and
+        #    np.any(left_hemi[int(x-10)]):
+        #        pass
+        # Case 3: Two Connected component in both hemispheres
```
Comment (GaelVaroquaux) [inline nilearn/plotting/find_cuts.py]: The code above needs to be cleaned up: the if clause should be removed and the else should be turned to a "if not ()" the clause above.

File: nilearn/plotting/tests/test_find_cuts.py (hunk 2 above)
Comment (GaelVaroquaux) [reply nilearn/plotting/tests/test_find_cuts.py]: I agree that we should check that the code doesn't crash when one map is empty.

**Comment by GaelVaroquaux** [comment]
We should also reference these functions in the plotting documentation, next to the plot_connectome documentation.

That's all I have. This is almost ready!

File: examples/03_connectivity/plot_atlas_comparison.py (hunk 7 above)
Comment (KamalakerDadi) [reply examples/03_connectivity/plot_atlas_comparison.py]: Done. See below.

File: nilearn/plotting/tests/test_find_cuts.py (hunk 2 above)
Comment (KamalakerDadi) [reply nilearn/plotting/tests/test_find_cuts.py]: >I agree that we should check that the code doesn't crash when one map is empty.

Code is not crashing. But, coordinates returned on an empty map depends on the shape of the map. Example: An empty map of shape=(50, 50, 50) is given, coords are returned as (25, 25, 25).

This is find_xyz_cut_coords specific. Should I open a new PR to raise a warning when an empty map is given and return coords=[(0, 0, 0)] ?

Comment (KamalakerDadi) [reply nilearn/plotting/tests/test_find_cuts.py]: I made a PR for this #1707

File: doc/connectivity/functional_connectomes.rst [hunk 9]
```diff
@@ -205,29 +205,37 @@ In the case of the MSDL atlas
...
-(see example:
-:ref:`sphx_glr_auto_examples_03_connectivity_plot_multi_subject_connectome.py`)::
+As you can see, the correlation matrix gives a very "full" graph: every
+node is connected to every other one. This is because it also captures
+indirect connections. In the next section we will see how to focus on
+only direct connections.
 
- >>> from nilearn import image, plotting
- >>> atlas_region_coords = [plotting.find_xyz_cut_coords(img) for img in image.iter_img(atlas_filename)] # doctest: +SKIP
+A functional connectome: extraction of MNI coordinates
```
Comment (GaelVaroquaux) [inline doc/connectivity/functional_connectomes.rst]: I would say "A functional connectome: extracting coordinates of regions"

File: examples/03_connectivity/plot_atlas_comparison.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
+"""
+Comparing connectomes on different reference atlases
+====================================================
+
+An important part of the example, is to turn a parcellation into
```
Comment (GaelVaroquaux) [inline examples/03_connectivity/plot_atlas_comparison.py]: I would rephrase this as follows: "Shows how to turn a parcellation into a connectome for visualization. ..."

File: nilearn/plotting/find_cuts.py [hunk 11]
```diff
@@ -337,3 +337,115 @@ def find_cut_slices(img, direction='z', n_cuts=7, spacing='auto'):
...
+        Label regions cut coordinates in image space (mm).
+
+    labels_list: list, optional
+        Label region. Returned only when return_label_names is True.
+
+    See Also
+    --------
+    nilearn.plotting.find_probabilistic_atlas_cut_coords : For coordinates
+        extraction on probabilistic atlases (4D) (Eg. MSDL atlas)
+    """
```
Comment (GaelVaroquaux) [inline nilearn/plotting/find_cuts.py]: We should check that label_hemisphere is in 'right', 'left', and raise a meaningful error here.

File: examples/03_connectivity/plot_atlas_comparison.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+given directly to :class:`nilearn.connectome.ConnectivityMeasure` for
+computing parcel-wise correlation matrices for each atlas across all subjects.
+
+Mean correlation matrix is displayed on glass brain on extracted coordinates.
+
+# author: Amadeus Kanaan
+
+"""
+# General imports
+import numpy as np
```
Comment (GaelVaroquaux) [inline examples/03_connectivity/plot_atlas_comparison.py]: I don't see where this is used (maybe I didn't look well)

File: examples/03_connectivity/plot_atlas_comparison.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+print('Yeo atlas nifti image (3D) with 17 parcels and liberal mask is located '
+      'at: %s' % yeo['thick_17'])
+print('Harvard Oxford atlas nifti image (3D) thresholded at .25 is located '
+      'at: %s' % harvard_oxford['maps'])
+# Store atlases in a dictionary. Atlases are for extracting coordinates.
+atlases = {
+    'Destrieux Atlas (struct)': destrieux['maps'],
+    'Yeo Atlas 17 thick (func)': yeo['thick_17'],
+    'Harvard Oxford > 25% (struct one hemisphere)': harvard_oxford['maps'],
+    'Harvard Oxford > 25% (struct two hemispheres)': harvard_oxford_s['maps']}
```
Comment (GaelVaroquaux) [inline examples/03_connectivity/plot_atlas_comparison.py]: That one gives a connectome that is way too dense, and hard to look at. I think that we should remove it.

Comment (GaelVaroquaux) [reply examples/03_connectivity/plot_atlas_comparison.py]: We should probably also remove the Destrieux: that script takes 9mn to run on the CI. That's a lot.

Comment (GaelVaroquaux) [reply examples/03_connectivity/plot_atlas_comparison.py]: Actually, thinking of it, I am in favor of simplifying that script down all the way to having only 1 atlas of each kind. It is enough to convey the message. I would keep Yeo and MSDL.

**Comment by KamalakerDadi** [comment]
>Maybe we could add the function also in plot_extract_regions_dictlearning_maps

In this example, at some point cut_coords are extracted on a **single map** using find_xyz_cut_coords. Whereas this function will accept multiple maps interated over each map checked by ```check_niimg_4d```.

**Comment by KamalakerDadi** [comment]
>Maybe we could add the function also in plot_extract_regions_dictlearning_maps

>In this example, at some point cut_coords are extracted on a single map using find_xyz_cut_coords. >Whereas this function will accept multiple maps interated over each map checked by check_niimg_4d.

@GaelVaroquaux what is your inputs here ? Does this needs to be worked on a images which has single map (3D) ?

**Comment by GaelVaroquaux** [comment]
>     Maybe we could add the function also in
>     plot_extract_regions_dictlearning_maps

>   In this example, at some point cut_coords are extracted on a single map
>   using find_xyz_cut_coords. >Whereas this function will accept multiple maps
>   interated over each map checked by check_niimg_4d.

> @GaelVaroquaux what is your inputs here ? Does this needs to be worked on a
> images which has single map (3D) ?

In 8.4.13.5. it's used on a 4D image:
http://nilearn.github.io/auto_examples/03_connectivity/plot_extract_regions_dictlearning_maps.html#plot-resulting-connectomes

I was thinking to add it there. It would simplify by avoiding the need
for iter_img.

**Comment by GaelVaroquaux** [comment]
AppVeyor is not important: it's the download failure.

**Comment by GaelVaroquaux** [comment]
LGTM. Merging. Thank you!

**Comment by abozoki** [comment]
Hello, I'm looking to obtain the center-of-mass coordinates for every ROI in the standard Freesurfer 2009b parcellation atlas in order to do nodal connectivity analysis. It seems that your program would do that, but I can't help wondering if this output doesn't already exist. If so, could I obtain that data from you or do you know who might have it?

Thank you in advance,
Andrea Bozoki, MD
Professor of Neurology
Michigan State University
